astropy/coordinates/spectral_coordinate.py

```diff
@@ -0,0 +1,851 @@
+                             "observer.")
+
+        target = self.target if target is None else target
+
+        # Check velocities and distance values on the new frame. This is
+        #  handled in the frame validation.
+        observer = self._validate_coordinate(observer)
+
+        # Calculate the initial and final los velocity
+        init_obs_vel = self._calculate_radial_velocity(self.observer, target)
+        fin_obs_vel = self._calculate_radial_velocity(observer, target)
+
+        line_of_sight_unit_vec = self._norm_d_pos(observer, target).xyz
+
+        new_data = self._project_velocity_and_shift(init_obs_vel, fin_obs_vel, line_of_sight_unit_vec)
+
+        new_coord = self._copy(value=new_data,
+                               observer=observer,
+                               target=target)
+
+        return new_coord
+
+    def _project_velocity_and_shift(self, init_vel, fin_vel, line_of_sight_unit_vec):
+        """
+        Calculated the velocity projection given two vectors.
+
+        Parameters
+        ----------
+        init_vel : `u.Quantity`
+            Initial velocity vector.
+        fin_vel : `u.Quantity`
+            Final velocity vector.
+
+        Returns
+        -------
+        new_data : `u.Quantity`
+            Spectral axis data with velocity shift applied.
+        """
+
+        # Project the velocity shift vector onto the line-of-sight vector
+        # between the target and the new observation frame.
+        init_proj_vel = np.dot(init_vel, line_of_sight_unit_vec) * line_of_sight_unit_vec
+
+        if NUMPY_LT_1_17:
+            init_proj_vel *= init_vel.unit
+
+        # Calculate the magnitude of the velocity shift between the two vectors.
+        # The vectors are aligned but may be in opposite directions, so we use
+        # the dot product to determine this.
+
+        diff_vel = fin_vel - init_proj_vel
+        delta_vel = np.dot(diff_vel, line_of_sight_unit_vec)
+
+        if NUMPY_LT_1_17:
+            delta_vel *= diff_vel.unit
+
+        # In the case where the projected velocity is nan, we can assume that
+        # the final velocity different is zero, and thus the actual velocity
+        # delta is equal to the original radial velocity.
+
+        # TODO: Due to lack of precision in some coordinate transformations,
+        # we may end up with a final velocity very close to, but not quite at,
+        # zero. In this case, set a tolerance for the final velocity; if it's
+        # below this tolerance, assume the delta velocity is essentially nan.
+
+        fin_vel_mag = np.linalg.norm(fin_vel)
+
+        if NUMPY_LT_1_17:
+            fin_vel_mag *= fin_vel.unit
+
+        if np.isnan(delta_vel) or fin_vel_mag < 1e-7 * fin_vel.unit:
+            delta_vel = -self.radial_velocity
+
+        return self._apply_relativistic_doppler_shift(delta_vel)
+
+    def with_observer_velocity(self, frame, preserve_observer_frame=False):
```

Comment:
Hm, this name is confusing to me, but I don't immediately have a better suggestion...

Reply:
Ha, this has been the subject of many discussions so far and I've not yet found a name that satisfies everyone! Other names it has had though the commit history are: ``in_observer_velocity_frame``, ``with_observer_in_velocity_frame_of``, and I've also considered ``with_observer_stationary_relative_to``. Naming things is hard :stuck_out_tongue:

Reply:
My suggestion (see earlier) is `transform_observer_to()`, though now that I look at it more closely, perhaps it doesn't make so much sense. The more general `transform_to` might still work, though, and is nicely analogous to `SkyCoord`.

Reply:
The reason I originally suggested something like the current name is that this routine is leaving the position of the observer unchanged, but changing the observers velocity *and* the frame in which it is represented. 

Suggestions along the `transform*` route suggest the observer remains unchanged which is misleading. 

I also suggested not changing the observer’s frame here but I think the objection is that it produces output that is correct but surprising (v large ITRS velocities)

Reply:
Yikes, I'm seeing how this is indeed tricky.

Reply:
Of the many suggestions so far I think `with_observer_stationary_relative_to` is verbose but clearest. Maybe just `stationary_relative_to` would be OK?

I have a suggestion - not sure how big a change it would be. The idea is to make it clearer what is happening by changing the arguments. So instead of passing in a frame instance whose position argument is (perhaps confusingly) ignored, pass the velocity and frame *type* separately.

```python
sc_tttau_bary = sc_ttau.stationary_relative_to('icrs', velocity=vel)
sc_tttau_bary = sc_ttau.stationary_relative_to(coord.ICRS, vel)
```

The velocity argument would be anything that can be converted to a ```CartesianDifferential``` and would be *optional* so:

```python
sc_tttau_bary = sc_ttau.stationary_relative_to('icrs')
```

would give a ```SpectralCoord``` with an observer stationary w.r.t to the barycenter. 
This looks cleaner and more intuitive to my eyes, but the downside would be that getting the ```SpectralCoord``` relative to the rest frame of the target would be a lot clunkier:

```python
sc_ttau_rest = sc_ttau.stationary_relative_to(
    sc_ttau.target.frame, 
    sc_ttau.target.cartesian.differentials['s']
)
```

Reply:
Thanks for the new idea! See my latest comment on the PR.